q2_diversity_lib/_util.py

```diff
@@ -81,21 +81,19 @@ def _validate_requested_cpus(wrapped_function, *args, **kwargs):
     # If `Process.cpu_affinity` unavailable on system, fall back
     # https://psutil.readthedocs.io/en/latest/index.html#psutil.cpu_count
     try:
-        cpus = len(psutil.Process().cpu_affinity())
+        cpus_available = len(psutil.Process().cpu_affinity())
     except AttributeError:
-        cpus = psutil.cpu_count(logical=False)
+        cpus_available = psutil.cpu_count(logical=False)
 
-    if isinstance(cpus_requested, int) and cpus_requested > cpus:
+    if isinstance(cpus_requested, int) and cpus_requested > cpus_available:
         raise ValueError(f"The value passed to '{param_name}' cannot exceed "
-                         f"the number of processors ({cpus}) available to "
-                         "the system.")
+                         f"the number of processors ({cpus_available}) "
+                         "available to the system.")
 
     if cpus_requested == 'auto':
-        # remove 'auto' from args to prevent 'multiple values' TypeError...
-        argslist = list(args)
-        argslist.remove('auto')
-        return_args = tuple(argslist)
-        # ...then inject number of available cpus
-        return wrapped_function(*return_args, **kwargs, **{param_name: cpus})
+        # replace 'auto' with the requested number of cpus and return
+        b_a_arguments[param_name] = cpus_available
+        return wrapped_function(*bound_arguments.args,
```

Comment:
When you move this recast up, drop the early return (this line) - you can return once, on line 99. It'll be easier to reason about for future readers (I'm not opposed to early returns, in general, but this one will just make the code more complicated.

Reply:
Doing this without an `isinstance` check causes `TypeError`s when 'auto' is passed, unless I also update `cpus_requested` to match `b_a_arguments[param_name]` a la:

![image](https://user-images.githubusercontent.com/39198770/84436899-09e5b480-abe9-11ea-8786-4a3aab1f9f1f.png)

With this in mind, the early return feels cleaner to me. I have retained it for now, as the return fully decouples the two checks. If `cpus_requested == auto`, we mutate and return. If not, we check for appropriate number of jobs and return. I really respect your judgement here, though, and would be happy to switch this over to a single return as in the screencap if you still feel that's more readable.

Reply:
> unless I also update cpus_requested to match b_a_arguments[param_name]

Yeah, that was [my point](https://github.com/qiime2/q2-diversity-lib/pull/18#discussion_r439034432) - the `cpus_requested > cpus_available` check should be handed the updated `cpus_requested` value (after casting from auto->int), as a last ditch check, to make sure everything is squeaky clean. Sorry for not making that clear.

> single return as in the screencap if you still feel that's more readable.

Its not so much that it isn't readable - its that you were maintaining two separate groups of state (`bound_arguments` vs `args & kwargs`). That is error prone and introduces more overhead for future readers of the code - I think that it unnecessarily complicates things.

What really matters here to me is that you only really need to return one set of arguments: `*bound_arguments.args, **bound_arguments.kwargs`. In your latest round of updates I see you're returning the exact same thing in both return statements (which is good), but, I think that makes the first `return` redundant, which might be confusing to future readers. As well, I think for the sake of consistency (and future comprehension), applying the same set of checks to the `cpus_requested` variable is important - it decreases the complexity of the code and makes it less likely that someone will introduce an accidental bug in the future. If we think about it in terms of data-flow, its almost always better to prevent branches - branches represent departures in state that require us to mentally track. In this case, the early return is a branch.